AI behaviour change

EmperorPinguin opened the pull request:
First time adding code to the game, please look for errors before merging. It's hard to determine if the AI actually becomes stronger, but it does appear to play a bit more sensible with the changes.

EmperorPinguin commented on core/src/com/unciv/logic/automation/Automation.kt:
```diff
@@ -91,16 +94,29 @@ object Automation {
-                yieldStats.culture *= 2
+            if (city.civ.stats.statsForNextTurn.gold <= 0 && city.civ.getHappiness() > -1)
+                yieldStats.food *= 2 // Try to increase our trade route gold
+                yieldStats.food *= 2
```
The second food is meant to be a gold. I might have made a bug there, let me check Kotlin documentation.

EmperorPinguin commented:
> Best explain the rationale for removing `if (city.tiles.size < 12) yieldStats.culture *= 2`. Otherwise, I see a lot of tweaks that I'd read "that guy surely has tweaked and fiddled and game-tested _a lot_ to arrive there", so I'm not going to analyze. That's more @tuvus 's domain.

I did increase the weight for constructing culture buildings in ConstructionAutomation.kt, as the AI was delaying their monuments for too long.  I don't get the rationale for making the AI work more culture when the city is small, you don't want the AI to resort to working artist slots by accident. I didn't obsevere weird effects when removing it. Is it to prioritise culture pantheon tiles (e.g. 2 food 1 culture vs 2 food 1 gold)? This should apply throughout the whole game I think, not just for cities with low amount of tiles.

tuvus commented:
From what I can tell, you are trying to optimize certain choices the AI makes by changing what it values. The problem is that you are making changes on a global scale that will last beyond the one decision you are trying to influence.

For example, your changes say that cities should, in almost every case, value science 3 times more and gold half as much just because we have tech to research and a larger city.

You also make other changes like increasing the value of spaceship parts and being afraid of barbarians, when the solution likely lies in the way the AI researches technology or looks for the barbarians. It's important to know that changing the values to be higher won't efficiently solve every problem.

As for the CityLocationTileRanker, I don't understand why you have made those changes.

> `if (city.tiles.size < 12) yieldStats.culture *= 2`

By the way, your PR is making the AI more likely to work artist slots since you have moved the line of code outside of the `        if (areWeRankingSpecialist) {} else {` branch

EmperorPinguin commented:
> For example, your changes say that cities should, in almost every case, value science 3 times more and gold half as much just because we have tech to research and a larger city.
Human players will lock scientist slots, but avoid merchants and most other specialists. The way to let the AI behave like this is by changing values like this. It's probably better do to this via an evaluation of different kinds of great person points though.
> You also make other changes like increasing the value of spaceship parts and being afraid of barbarians, when the solution likely lies in the way the AI researches technology or looks for the barbarians. It's important to know that changing the values to be higher won't efficiently solve every problem.
I've very aware of this, but I don't know how to make the AI prioritise techs or look better for barbarians. I'm first doing easier changes which do appear to have an effect in the right direction.
> As for the CityLocationTileRanker, I don't understand why you have made those changes.
I made comments for changes I added. 7 tiles away is too much as the road upkeep is higher, the roads are more expensive, it's slower to reach and harder to defend, and the AI often loses settlers trying to reach it. For the very late game, it's better to have cities far apart to work all tiles, but games are already decided then. Hills and rivers are good, that's why I increased the value for settling on them. 
> > `if (city.tiles.size < 12) yieldStats.culture *= 2`
I reverted the change.
